### PR TITLE
wallet: Migrate entire address book entries to watchonly and solvables too

### PR DESCRIPTION
**Pull request by achow101**
Not all of the data in an address book entry was being copied to the watchonly and solvables wallets. This includes information such as whether the address was previously spent, and any receive requests that may exist. A test has been added to check that the previously spent information is copied, although it passes without the changes in this PR since this information is also regenerated when a transaction is loaded/added into a wallet.

### CODE REVIEW

File: test/functional/wallet_migration.py [hunk 1]
```diff
@@ -878,6 +878,63 @@ def test_failed_migration_cleanup(self):
...
+
+        wallet = self.create_legacy_wallet("avoidreuse")
+        wallet.setwalletflag("avoid_reuse", True)
+
+        # Import a pubkey to the test wallet and send some funds to it
+        reused_imported_addr = def_wallet.getnewaddress()
+        wallet.importpubkey(def_wallet.getaddressinfo(reused_imported_addr)["pubkey"])
+        txid = def_wallet.sendtoaddress(reused_imported_addr, 2)
+        vout = find_vout_for_address(def_wallet, txid, reused_imported_addr)
+        imported_utxo = {"txid": txid, "vout": vout}
```
[the remark this answers is not in the document]
Comment (achow101) [reply test/functional/wallet_migration.py]: used the new helper.

File: src/wallet/wallet.cpp [hunk 2]
```diff
@@ -3960,25 +3960,17 @@ bool CWallet::ApplyMigrationData(MigrationData& data, bilingual_str& error)
...
                     LOCK(data.watchonly_wallet->cs_wallet);
                     if (data.watchonly_wallet->IsMine(addr_pair.first)) {
-                        // Add to the watchonly. Preserve the labels, purpose, and change-ness
-                        std::string label = addr_pair.second.GetLabel();
-                        data.watchonly_wallet->m_address_book[addr_pair.first].purpose = addr_pair.second.purpose;
-                        if (!addr_pair.second.IsChange()) {
-                            data.watchonly_wallet->m_address_book[addr_pair.first].SetLabel(label);
-                        }
+                        // Add to the watchonly. Copy the entire address book entry
+                        data.watchonly_wallet->m_address_book[addr_pair.first] = addr_pair.second;
```
Comment (ryanofsky) [inline src/wallet/wallet.cpp:3964]: In commit "wallet: Migrate entire address book entries" (02665016a5870c9fcba049b849a4086b14d08255)

This is so much more straightforward, thank you for cleaning it up. But while you are getting rid of confusing GetLabel code here, I think you should do it below too:

```diff
--- a/src/wallet/wallet.cpp
+++ b/src/wallet/wallet.cpp
@@ -4007,10 +4007,8 @@ bool CWallet::ApplyMigrationData(MigrationData& data, bilingual_str& error)
         WalletBatch batch{wallet.GetDatabase()};
         for (const auto& [destination, addr_book_data] : wallet.m_address_book) {
             auto address{EncodeDestination(destination)};
-            std::optional<std::string> label = addr_book_data.IsChange() ? std::nullopt : std::make_optional(addr_book_data.GetLabel());
-            // don't bother writing default values (unknown purpose)
             if (addr_book_data.purpose) batch.WritePurpose(address, PurposeToString(*addr_book_data.purpose));
-            if (label) batch.WriteName(address, *label);
+            if (addr_book_data.label) batch.WriteName(address, *addr_book_data.label);
             for (const auto& [id, request] : addr_book_data.receive_requests) {
                 batch.WriteAddressReceiveRequest(destination, id, request);
             }

```

**Review by ryanofsky — APPROVED** [review]
Code review ACK 02665016a5870c9fcba049b849a4086b14d08255. Nice cleanup and test!

**Review by ryanofsky — APPROVED** [review]
Code review ACK 406b71abcb72f234ddf9245a3f57e748343c774f. Just suggested change since last review

**Review by furszy — APPROVED** [review]
Code review ACK 406b71ab


With two non-blocking points:

Firstly, made a test to verify that the transaction's extra information is preserved during migration: c1aabd1b2a. This includes the preservation of bump fee info ('replaces_txid' and 'replaced_by_txid') and the user's custom comments ('comment' and 'comment_to'). Feel free to cherry-pick it, or I can push it in a follow-up. Either way is fine for me.

Secondly, shilling mode; with #26836, which now is part of #28574, this could have been implemented with even less code.

**Pull request merged**